miden-tx/src/host/event/vault_delta.rs

```diff
@@ -0,0 +1,165 @@
+use miden_objects::{
+    accounts::{AccountId, AccountVaultDelta},
+    assets::{Asset, FungibleAsset, NonFungibleAsset},
+    utils::collections::{btree_map::Entry, BTreeMap},
+    Digest,
+};
+use vm_processor::{ExecutionError, HostResponse, ProcessState};
+
+/// The [AccountVaultDeltaHandler] is responsible for tracking changes to the vault of the account
+/// the transaction is being executed against.
+///
+/// It is composed of two maps:
+/// - [AccountVaultDeltaHandler::fungible_assets] - tracks changes to the vault's fungible assets,
+/// where the key is the faucet ID of the asset, and the value is the amount of the asset being
+/// added or removed from the vault.
+/// - [AccountVaultDeltaHandler::non_fungible_assets] - tracks changes to the vault's non-fungible
+/// assets, where the key is the non-fungible asset, and the value is either 1 or -1 depending
+/// on whether the asset is being added or removed from the vault.
+#[derive(Default, Debug)]
+pub struct AccountVaultDeltaHandler {
+    fungible_assets: BTreeMap<u64, i128>,
+    non_fungible_assets: BTreeMap<Digest, i8>,
+}
+
+impl AccountVaultDeltaHandler {
+    // MODIFIERS
+    // --------------------------------------------------------------------------------------------
+
+    /// Extracts the asset that is being added to the account's vault from the process state and
+    /// updates the appropriate [AccountVaultDeltaHandler::fungible_assets] or
+    /// [AccountVaultDeltaHandler::non_fungible_assets] map.
+    pub fn add_asset<S: ProcessState>(
+        &mut self,
+        process: &S,
+    ) -> Result<HostResponse, ExecutionError> {
+        let asset: Asset = process.get_stack_word(0).try_into().map_err(|err| {
+            ExecutionError::EventError(format!(
+                "Failed to apply account vault delta - asset is baldy formed - {err}"
+            ))
+        })?;
+
+        match asset {
+            Asset::Fungible(asset) => {
+                update_asset_delta(
+                    &mut self.fungible_assets,
+                    asset.faucet_id().into(),
+                    asset.amount() as i128,
+                );
+            }
+            Asset::NonFungible(asset) => {
+                update_asset_delta(&mut self.non_fungible_assets, asset.vault_key().into(), 1)
```

Comment:
Using $1$ and $-1$ for non-fungible asset counts could technically allow incorrect intermediate state - e.g., if an asset is added twice and then removed once. This should be caught by the kernel code - but I wonder if we should try to catch this here as well.

Reply:
I don't think it should be the responsibility of the `AccountVaultDeltaHandler` to evaluate the correctness of `AccountVault` state transitions, that should be the responsibility of the kernel.  The kernel will panic if such an invalid transition is executed - https://github.com/0xPolygonMiden/miden-base/blob/d92549b037918a6ca08356794ecfe3dd66367a91/miden-lib/src/tests/test_asset_vault.rs#L251-L288.

Reply:
Yep - I think that's fine.